src/main/java/org/craftercms/studio/impl/v2/service/security/UserServiceImpl.java

```diff
@@ -151,8 +156,46 @@ public void updateUser(User user) throws ServiceLayerException {
 
     @Override
     @HasPermission(type = DefaultPermission.class, action = "delete_users")
-    public void deleteUsers(List<Long> userIds, List<String> usernames) throws ServiceLayerException {
-        securityProvider.deleteUsers(userIds, usernames);
+    public void deleteUsers(List<Long> userIds, List<String> usernames) throws ServiceLayerException, AuthenticationException {
+        User currentUser = getCurrentUser();
+
+        if (CollectionUtils.containsAny(userIds, Arrays.asList(currentUser.getId())) ||
+                CollectionUtils.containsAny(usernames, Arrays.asList(currentUser.getUsername()))) {
+            throw new ServiceLayerException("Cannot delete self.");
+        }
+
+        generalLockService.lock(REMOVE_SYSTEM_ADMIN_MEMBER_LOCK);
+        try {
+            try {
+                Group g = groupService.getGroupByName(SYSTEM_ADMIN_GROUP);
+                List<User> members = groupService.getGroupMembers(g.getId(), 0, Integer.MAX_VALUE, StringUtils.EMPTY);
+                if (CollectionUtils.isNotEmpty(members)) {
+                    List<User> membersAfterRemove = new ArrayList<User>();
+                    membersAfterRemove.addAll(members);
+                    members.forEach(m -> {
+                        if (CollectionUtils.isNotEmpty(userIds)) {
+                            if (userIds.contains(m.getId())) {
+                                membersAfterRemove.remove(m);
+                            }
+                        }
+                        if (CollectionUtils.isNotEmpty(usernames)) {
+                            if (usernames.contains(m.getUsername())) {
+                                membersAfterRemove.remove(m);
+                            }
+                        }
+                    });
+                    if (CollectionUtils.isEmpty(membersAfterRemove)) {
+                        throw new ServiceLayerException("Removing all users from System Admin group is not allowed");
```

Comment:
Change the wording to: "Removing all members of the System Admin group is not allowed. We must have at least one system administrator."